codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/Waiters.java

```diff
@@ -0,0 +1,690 @@
+        );
+        writer.openBlock(
+                "func (w $P) $L(ctx context.Context, params $P, maxWaitDur time.Duration, optFns ...func($P)) error {",
+                "}",
+                clientSymbol, WAITER_INVOKER_FUNCTION_NAME, inputSymbol, waiterOptionsSymbol,
+                () -> {
+                    writer.openBlock("if maxWaitDur == 0 {", "}", () -> {
+                        writer.addUseImports(SmithyGoDependency.FMT);
+                        writer.write("fmt.Errorf(\"maximum wait time for waiter must be greater than zero\")");
+                    }).write("");
+
+                    writer.write("options := w.options");
+
+                    writer.openBlock("for _, fn := range optFns {",
+                            "}", () -> {
+                                writer.write("fn(&options)");
+                            });
+                    writer.write("");
+
+                    // validate values for MaxDelay from options
+                    writer.openBlock("if options.MaxDelay == 0 {", "}", () -> {
+                        writer.write("options.MaxDelay = $L * time.Second", waiter.getMaxDelay());
+                    });
+                    writer.write("");
+
+                    // validate that MinDelay is lesser than or equal to resolved MaxDelay
+                    writer.openBlock("if options.MinDelay > options.MaxDelay {", "}", () -> {
+                        writer.addUseImports(SmithyGoDependency.FMT);
+                        writer.write("return fmt.Errorf(\"minimum waiter delay %v must be lesser than or equal to "
+                                + "maximum waiter delay of %v.\", options.MinDelay, options.MaxDelay)");
+                    }).write("");
+
+
+                    writer.addUseImports(SmithyGoDependency.CONTEXT);
+                    writer.write("ctx, cancelFn := context.WithTimeout(ctx, maxWaitDur)");
+                    writer.write("defer cancelFn()");
+                    writer.write("");
+
+                    Symbol loggerMiddleware = SymbolUtils.createValueSymbolBuilder(
+                            "Logger", SmithyGoDependency.SMITHY_WAITERS
+                    ).build();
+                    writer.write("logger := $T{}", loggerMiddleware);
+                    writer.write("remainingTime := maxWaitDur").write("");
+
+                    writer.write("var attempt int64");
+                    writer.openBlock("for {", "}", () -> {
+                        writer.write("");
+
+                        // handle retry delay computation, sleep. Skip if minDelay is lesser than or equal to 0.
+                        writer.openBlock("if attempt > 0 && options.MinDelay > 0 {", "}", () -> {
+                            writer.openBlock("if remainingTime < options.MinDelay || remainingTime <= 0 {", "}", () -> {
+                                writer.write("break");
+                            });
+                            writer.write("");
```

Comment:
kind of an edge case here, but if `MinDelay` is `0`, then i think this will loop forever, until the context is canceled. since `remainingTime` is only decremented based on delay.  To address this, consider moving `remainingTime` check a level higher

In addition, should also decrement `remainingTime` based on the duration the request took to execute, in addition to the `delay` sleep for.  Without this `remainingTime` only tracks delay decrements not the time the operations took in addition. Which means that the context timeout would also always trigger before the remainingTime is < minDelay.  Ideally we'd prefer the remainingTime < minDelay to be the case that causes the waiter to return.